File: MiniDOMBase/DMD/Sistema/Repositories/CGeneralClass.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.internals;

namespace minidom
{

    namespace repositories
    {

        /// <summary>
        /// Handler degli eventi
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        public delegate void ModuleEventHandler(object sender, EventDescription e);

        /// <summary>
        /// Gestore delle risorse
        /// </summary>
        [Serializable]
        public abstract class CModulesClass 
            : IEnumerable
        {


            /// <summary>
            /// Notifica al sistema l'inserimento di un nuovo oggetto
            /// </summary>
            /// <remarks></remarks>
            public event ItemCreatedEventHandler ItemCreated;

            

            /// <summary>
            /// Notifica al sistema l'eliminazione di un oggetto
            /// </summary>
            /// <remarks></remarks>
            public event ItemDeletedEventHandler ItemDeleted;

            

            /// <summary>
            /// Notifica al sistema che un oggetto è stato modificato
            /// </summary>
            /// <remarks></remarks>
            public event ItemModifiedEventHandler ItemModified;

            /// <summary>
            /// Evento genirico sul modulo
            /// </summary>
            public event ModuleEventHandler ModuleEvent;

            private string m_ModuleName;
            private Type m_CursorType;
            [NonSerialized] private CModule m_Module;
            [NonSerialized] private List<CacheItem> m_CachedItems;
            private int m_CacheSize;
            [NonSerialized] private CCollection m_AllItems;
            private bool m_isInit;
            [NonSerialized] private DBConnection m_Conn = null;

            [NonSerialized] private CIndexingService m_Index = null;

            /// <summary>
            /// Oggetto utilizzabile per la sincronizzazione del singolo modulo
            /// </summary>
            /// <remarks></remarks>
            [NonSerialized] public readonly object cacheLock = new object();


            /// <summary>
            /// Elemento nella cache
            /// </summary>
            [Serializable]
            protected class CacheItem
            {
                /// <summary>
                /// Elmento
                /// </summary>
                public object Item;

                /// <summary>
                /// Vita
                /// </summary>
                public int Live;

                /// <summary>
                /// Costruttore
                /// </summary>
                public CacheItem()
                {
                    ////DMDObject.IncreaseCounter(this);
                }

                /// <summary>
                /// Costruttore
                /// </summary>
                /// <param name="item"></param>
                public CacheItem(object item)
                {
                    ////DMDObject.IncreaseCounter(this);
                    Item = item;
                    Live = 1;
                }
 
            }


            /// <summary>
            /// Costruttore
            /// </summary>
            public CModulesClass()
            {
                ////DMDObject.IncreaseCounter(this);
                m_ModuleName = "";
                m_CursorType = null;
                m_Module = null;
                m_CacheSize = 0;
                m_CachedItems = new List<CacheItem>();
                m_AllItems = null;
                m_isInit = false;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="moduleName"></param>
            public CModulesClass(string moduleName) 
                : this()
            {
                moduleName = DMD.Strings.Trim(moduleName);
                if (DMD.Strings.Left(moduleName, 3) != "mod")
                    throw new ArgumentNullException("moduleName");
                if (string.IsNullOrEmpty(moduleName))
                    throw new ArgumentNullException("moduleName");
                m_ModuleName = moduleName;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="moduleName"></param>
            /// <param name="cursorType"></param>
            public CModulesClass(string moduleName, Type cursorType) 
                : this()
            {
                moduleName = DMD.Strings.Trim(moduleName);
                if (DMD.Strings.Left(moduleName, 3) != "mod")
                    throw new ArgumentNullException("moduleName");
                if (string.IsNullOrEmpty(moduleName))
                    throw new ArgumentNullException("moduleName");
                if (cursorType is null)
                    throw new ArgumentNullException("cursorType");
                m_ModuleName = moduleName;
                m_CursorType = cursorType;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="moduleName"></param>
            /// <param name="cursorType"></param>
            /// <param name="cacheSize"></param>
            public CModulesClass(string moduleName, Type cursorType, int cacheSize)
                : this()
            {
                moduleName = DMD.Strings.Trim(moduleName);
                if (DMD.Strings.Left(moduleName, 3) != "mod")
                    throw new ArgumentNullException("moduleName");
                if (string.IsNullOrEmpty(moduleName))
                    throw new ArgumentNullException("moduleName");
                if (cursorType is null)
                    throw new ArgumentNullException("cursorType");
                m_ModuleName = moduleName;
                m_CursorType = cursorType;
                m_Module = null;
                m_CacheSize = cacheSize;
            }

            /// <summary>
            /// Restituisce un riferimento al modulo
            /// </summary>
            /// <returns></returns>
            public CModule Module
            {
                get
                {
                    lock (this)
                    {
                        if (this.m_Module is null) this.m_Module = Modules.GetItemByName(m_ModuleName);
                        if (this.m_Module is null)
                        {
                            this.m_Module = CreateModuleInfo();
                            this.InitializeStandardActions();
                        }
                        return this.m_Module;
                    }
                }
            }

            /// <summary>
            /// Accede al sistema di indicizzazione
            /// </summary>
            public CIndexingService Index
            {
                get
                {
                    lock (this)
                    {
                        if (m_Index is null && !string.IsNullOrEmpty(this.m_ModuleName))
                        {
                            var db = this.Database;
                            var tblWordStats = this.m_ModuleName + "_IdxWStats";
                            var tblWordIndexes = this.m_ModuleName + "_IdxWIndex";
                            m_Index = new CIndexingService(db, tblWordStats, tblWordIndexes);
                            m_Index.Database = db;
                            m_Index.WordIndexFolder = System.IO.Path.Combine(Sistema.ApplicationContext.SystemDataFolder, this.m_ModuleName + "\\wordindex\\");
                            Sistema.FileSystem.CreateRecursiveFolder(m_Index.WordIndexFolder);
                        }

                        return m_Index;
                    }
                }
            }

            /// <summary>
            /// Restituisce o imposta la connessione al DB
            /// </summary>
            public DBConnection Database
            {
                get
                {
                    if (m_Conn is null)
                        return minidom.Sistema.ApplicationContext.APPConn;
                    return m_Conn;
                }

                set
                {
                    m_Conn = value;
                }
            }

            /// <summary>
            /// Inizializza le risorse necessarie al modulo
            /// </summary>
            public virtual void Initialize()
            {
                if (m_isInit)
                    return;
                var m = this.Module;
                m_isInit = true;
            }

            /// <summary>
            /// Crea un nuovo oggetto gestito dal modulo
            /// </summary>
            /// <param name="args"></param>
            /// <returns></returns>
            public virtual object InstantiateNewItem(object args)
            {
                using(var c = this.CreateCursor())
                {
                    return c.Add();
                }
            }

            /// <summary>
            /// Inizializza le azioni standard del modulo
            /// </summary>
            public virtual void InitializeStandardActions()
            {
                var m = this.Module;

                m.DefinedActions.EnsureAction("create");
                m.DefinedActions.EnsureAction("list");
                m.DefinedActions.EnsureAction("edit");
                m.DefinedActions.EnsureAction("delete");
                using (var c = this.CreateCursor())
                {
                    if (c is minidom.Databases.DBObjectCursor)
                    {
                        m.DefinedActions.EnsureAction("list_own");
                        m.DefinedActions.EnsureAction("edit_own");
                        m.DefinedActions.EnsureAction("delete_own");
                        m.DefinedActions.EnsureAction("list_assigned");
                        m.DefinedActions.EnsureAction("edit_assigned");
                        m.DefinedActions.EnsureAction("delete_assigned");
                    }
                    if (c is minidom.Databases.DBObjectCursorPO)
                    {
                        m.DefinedActions.EnsureAction("list_office");
                        m.DefinedActions.EnsureAction("edit_office");
                        m.DefinedActions.EnsureAction("delete_office");
                    }
                }

            }


            /// <summary>
            /// Rilascia le risorse necessarie al modulo
            /// </summary>
            public virtual void Terminate()
            {
                m_isInit = false;
            }

            /// <summary>
            /// Crea il modulo e salva le informazioni nel DB
            /// </summary>
            /// <returns></returns>
            protected virtual CModule CreateModuleInfo()
            {
                var ret = new CModule(m_ModuleName);
                ret.Stato = ObjectStatus.OBJECT_VALID;
                ret.Description = m_ModuleName;
                ret.DisplayName = m_ModuleName;
                ret.ClassHandler = "CBaseModuleHandler";
                ret.Save();
                return ret;
            }

            /// <summary>
            /// Crea un'istanza del cursore
            /// </summary>
            /// <returns></returns>
            public virtual Databases.DBObjectCursorBase CreateCursor()
            {
                return (Databases.DBObjectCursorBase)DMD.RunTime.CreateInstance(this.m_CursorType);
            }

            /// <summary>
            /// Restituisce l'elemento in base all'ID
            /// </summary>
            /// <param name="id"></param>
            /// <returns></returns>
            public virtual object GetItemById(int id)
            {
                lock (cacheLock)
                {
                    if (id == 0)
                        return null;
                    var ret = GetCachedItemById(id);
                    if (ret is null)
                    {
                        ret = GetNonCachedItemById(id);
                        if (ret is object)
                            AddToCache(ret);
                    }

                    return ret;
                }
            }

            /// <summary>
            /// Restituisce l'elemento in base all'ID
            /// </summary>
            /// <param name="id"></param>
            /// <returns></returns>
            protected virtual object GetCachedItemById(int id)
            {
                if (id == 0)
                    return null;
                for (int i = 0, loopTo = m_CachedItems.Count - 1; i <= loopTo; i++)
                {
                    var o = m_CachedItems[i];
                    if (DBUtils.GetID(o.Item, 0) == id)
                        return o.Item;
                }

                return null;
            }

            ///// <summary>
            ///// Restituisce un riferimento al database
            ///// </summary>
            ///// <returns></returns>
            //public virtual Databases.CDBConnection GetConnection()
            //{
            //    lock (this)
            //    {
            //        if (m_Conn is null)
            //        {
            //            var cursor = CreateCursor();
            //            if (cursor is object)
            //            {
            //                m_Conn = cursor.GetConnection();
            //                cursor.Dispose();
            //            }
            //        }

            //        return m_Conn;
            //    }
            //}

            /// <summary>
            /// Carica l'elemento dalla base di dati
            /// </summary>
            /// <param name="id"></param>
            /// <returns></returns>
            protected virtual object GetNonCachedItemById(int id)
            {
                if (id == 0)
                    return null;
                using (var cursor = this.CreateCursor())
                {
                    cursor.ID.Value = id;
                    cursor.IgnoreRights = true;
                    return cursor.Item;
                }
            }
                 

            /// <summary>
            /// Aggiunge l'elemento alla cache
            /// </summary>
            /// <param name="item"></param>
            protected virtual void AddToCache(object item)
            {
                if (m_CacheSize == 0)
                    return;
                m_CachedItems.Add(new CacheItem(item));
                if (m_CacheSize > 0)
                {
                    while (m_CachedItems.Count > m_CacheSize)
                        m_CachedItems.RemoveAt(0);
                }

                m_AllItems = null;
            }

            /// <summary>
            /// Aggiunge l'elemento alla cache
            /// </summary>
            /// <param name="item"></param>
            protected virtual void AddToCache1(object item)
            {
                if (m_CacheSize == 0)
                    return;
                m_CachedItems.Add(new CacheItem(item));
                if (m_CacheSize > 0)
                {
                    while (m_CachedItems.Count > m_CacheSize)
                        m_CachedItems.RemoveAt(0);
                }
            }

            /// <summary>
            /// Restituisce true se l'utente corrente è abilitato ad eseguire l'azione
            /// </summary>
            /// <param name="action"></param>
            /// <returns></returns>
            public virtual bool UserCanDoAction(CModuleAction action)
            {
                return action.UserCanDoAction(minidom.Sistema.Users.CurrentUser);
            }

            /// <summary>
            /// Restituisce true se l'utente corrente è abilitato ad eseguire l'azione
            /// </summary>
            /// <param name="actionName"></param>
            /// <returns></returns>
            public virtual bool UserCanDoAction(string actionName)
            {
                var a = this.Module.DefinedActions.GetItemByKey(actionName);
                if (a is null)
                    return false; // Throw New ArgumentOutOfRangeException("Il modulo non implementa l'azione [" & actionName & "]")
                return this.UserCanDoAction(a);
            }

            /// <summary>
            /// Restituisce true se l'utente corrente è abilitato ad eseguire l'azione
            /// </summary>
            /// <param name="actionID"></param>
            /// <returns></returns>
            public virtual bool UserCanDoAction(int actionID)
            {
                return this.UserCanDoAction(this.Module.DefinedActions.GetItemById(actionID));
            }


            /// <summary>
            /// Aggiorna la cache sulla base dello stato dell'oggetto
            /// </summary>
            /// <param name="item"></param>
            /// <remarks></remarks>
            public virtual void UpdateCached(object item)
            {
                lock (cacheLock)
                {
                    if (m_CacheSize == 0)
                        return;
                    for (int i = 0, loopTo = m_CachedItems.Count - 1; i <= loopTo; i++)
                    {
                        var o = m_CachedItems[i];
                        if (DBUtils.GetID(o.Item, 0) == DBUtils.GetID(item, 0))
                        {
                            //var oldItem = o.Item;
                            //oldItem.InitializeFrom(item);
                            //item = oldItem;
                            m_CachedItems.RemoveAt(i);
                            break;
                        }
                    }

                    if (item is Databases.IDBObject)
                    {
                        if (((Databases.IDBObject)item).Stato == ObjectStatus.OBJECT_VALID)
                        {
                            m_CachedItems.Add(new CacheItem(item));
                        }
                    }
                    else if (DBUtils.GetID(item, 0) != 0)
                    {
                        m_CachedItems.Add(new CacheItem(item));
                    }

                    if (m_CacheSize > 0)
                    {
                        while (m_CachedItems.Count > m_CacheSize)
                            m_CachedItems.RemoveAt(0);
                    }

                    // If (Me.m_AllItems IsNot Nothing) Then
                    // For i As Integer = 0 To Me.m_AllItems.Count - 1
                    // If (GetID(Me.m_AllItems(i)) = GetID(item)) Then
                    // Dim oldItem As T = Me.m_AllItems(i)
                    // oldItem.InitializeFrom(item)
                    // Me.m_CachedItems.RemoveAt(i)
                    // Exit For
                    // End If
                    // Next
                    // End If
                    m_AllItems = null;
                }
            }

            /// <summary>
            /// Rimuove l'elemento dalla cache
            /// </summary>
            /// <param name="item"></param>
            public virtual void RemoveCached(object item)
            {
                lock (cacheLock)
                {
                    if (m_CacheSize == 0)
                        return;
                    for (int i = 0, loopTo = m_CachedItems.Count - 1; i <= loopTo; i++)
                    {
                        var o = m_CachedItems[i];
                        if (DBUtils.GetID(o.Item, 0) == DBUtils.GetID(item, 0))
                        {
                            m_CachedItems.RemoveAt(i);
                            break;
                        }
                    }

                    m_AllItems = null;
                }
            }

            /// <summary>
            /// Carica in memoria tutti gli elementi gestiti dal modulo (se CacheSize è impostato a -1)
            /// </summary>
            /// <returns></returns>
            public virtual CCollection LoadAll()
            {
                lock (cacheLock)
                {
                    if (m_AllItems is object)
                        return m_AllItems;
                    if (m_CacheSize != 0)
                    {
                        using (var cursor = CreateCursor())
                        {
                            if (cursor is Databases.DBObjectCursor)
                            {
                                ((Databases.DBObjectCursor)cursor).Stato.Value = ObjectStatus.OBJECT_VALID;
                            }
                            cursor.IgnoreRights = true;
                            while (
                                      cursor.Read() 
                                   && (m_CacheSize < 0 || m_CachedItems.Count < m_CacheSize)
                                   )
                            {
                                var item1 = GetCachedItemById(DBUtils.GetID(cursor.Item, 0));
                                if (item1 is null)
                                    AddToCache1(cursor.Item);
                                // Me.UpdateCached(cursor.Item)
                                 
                            }

                        }
                    }

                    m_AllItems = new CCollection();
                    for (int i = 0, loopTo = m_CachedItems.Count - 1; i <= loopTo; i++)
                        m_AllItems.Add(m_CachedItems[i].Item);
                    //if (typeof(IComparable).IsAssignableFrom(typeof(T)))
                    //    m_AllItems.Sort();
                    return m_AllItems;
                }
            }

            /// <summary>
            /// Accede alla cache
            /// </summary>
            protected List<CacheItem> CachedItems
            {
                get
                {
                    return this.m_CachedItems;
                }
            }

            /// <summary>
            /// Genera l'evento ItemCreated
            /// </summary>
            /// <param name="e"></param>
            protected virtual void OnItemCreated(ItemEventArgs e)
            {
                this.UpdateCached(e.Item);
                ItemCreated?.Invoke(this, e);
            }

            /// <summary>
            /// Chiama il metodo OnItemCreated
            /// </summary>
            /// <param name="e"></param>
            internal void doOnItemCreated(ItemEventArgs e)
            {
                this.OnItemCreated(e);
            }

            /// <summary>
            /// Genera l'evento ItemDeleted
            /// </summary>
            /// <param name="e"></param>
            protected virtual void OnItemDeleted(ItemEventArgs e)
            {
                this.UpdateCached(e.Item);
                ItemDeleted?.Invoke(this, e);
            }

            /// <summary>
            /// Chiama il metodo OnItemDeleted
            /// </summary>
            /// <param name="e"></param>
            internal void doOnItemDeleted(ItemEventArgs e)
            {
                this.OnItemDeleted(e);
            }

            /// <summary>
            /// Genera l'evento ItemModified
            /// </summary>
            /// <param name="e"></param>
            protected virtual void OnItemModified(ItemEventArgs e)
            {
                this.UpdateCached(e.Item);
                ItemModified?.Invoke(this, e);
            }

            /// <summary>
            /// Chiama il metodo OnItemModified
            /// </summary>
            /// <param name="e"></param>
            internal void doOnItemModified(ItemEventArgs e)
            {
                this.OnItemModified(e);
            }

            /// <summary>
            /// Notifica un evento relativo al modulo
            /// </summary>
            /// <param name="e"></param>
            public virtual void DispatchEvent(EventDescription e)
            {
                e.SetModule(this.Module);
                this.ModuleEvent?.Invoke(this, e);
                minidom.Sistema.Events.DispatchEvent(e);
            }

            /// <summary>
            /// Restituisce un oggetto IEnumerator che consente di enumerare tutti gli oggetti
            /// gestiti dal repository
            /// </summary>
            /// <returns></returns>
            public IEnumerator GetEnumerator()
            {
                return this._GetEnumerator();
            }

            /// <summary>
            /// Restituisce un oggetto IEnumerator che consente di enumerare tutti gli oggetti
            /// gestiti dal repository
            /// </summary>
            /// <returns></returns>
            protected virtual IEnumerator _GetEnumerator()
            {
                if (this.m_CacheSize == -1)
                {
                    return this.LoadAll().GetEnumerator();
                }
                else
                {
                    return new MyCursorEnumerator(this);
                }
            }

            /// <summary>
            /// Enumeratore basato su cursore
            /// </summary>
            protected class MyCursorEnumerator
                : IEnumerator<object>
            {
                /// <summary>
                /// Owner
                /// </summary>
                public CModulesClass owner;

                /// <summary>
                /// Cursore
                /// </summary>
                public minidom.Databases.DBObjectCursorBase cursor;

                /// <summary>
                /// Costruttore
                /// </summary>
                /// <param name="owner"></param>
                public MyCursorEnumerator(CModulesClass owner)
                {
                    if (owner is null)
                        throw new ArgumentNullException("owner");
                    this.cursor = owner.CreateCursor();
                    if (this.cursor is minidom.Databases.DBObjectCursor)
                    {
                        ((minidom.Databases.DBObjectCursor)cursor).Stato.Value = ObjectStatus.OBJECT_VALID;
                    }
                }

                /// <summary>
                /// Elemento corrente
                /// </summary>
                public object Current
                {
                    get
                    {
                        return this.cursor.Item;
                    }
                }

                /// <summary>
                /// Rilascia le risorese
                /// </summary>
                public void Dispose()
                {
                    this.cursor.Dispose();
                    this.cursor = null;
                }

                /// <summary>
                /// Sposta all'elemento successivo
                /// </summary>
                /// <returns></returns>
                public bool MoveNext()
                {
                    return this.cursor.Read();
                }

                /// <summary>
                /// Resetta il cursore
                /// </summary>
                public void Reset()
                {
                    this.cursor.Clear();
                }
            }

        }
    }
}

File: MiniDOMBase/DMD/Sistema/CGeneralClass.vb
Imports System.Net.Mail 'importo il Namespace
Imports System.Net.Sockets
Imports minidom
Imports minidom.Databases

#Const UseGetNonCachedItemById = False

Partial Public Class Sistema

    Public MustInherit Class CModulesClass(Of T As DBObjectBase)


        ''' <summary>
        ''' Notifica al sistema l'inserimento di un nuovo oggetto
        ''' </summary>
        ''' <param name="sender"></param>
        ''' <param name="e"></param>
        ''' <remarks></remarks>
        Public Event ItemCreated(ByVal sender As Object, ByVal e As ItemEventArgs)

        ''' <summary>
        ''' Notifica al sistema l'eliminazione di un oggetto
        ''' </summary>
        ''' <param name="sender"></param>
        ''' <param name="e"></param>
        ''' <remarks></remarks>
        Public Event ItemDeleted(ByVal sender As Object, ByVal e As ItemEventArgs)

        ''' <summary>
        ''' Notifica al sistema che un oggetto è stato modificato
        ''' </summary>
        ''' <param name="sender"></param>
        ''' <param name="e"></param>
        ''' <remarks></remarks>
        Public Event ItemModified(ByVal sender As Object, ByVal e As ItemEventArgs)



        Private m_ModuleName As String
        Private m_CursorType As System.Type
        Private m_Module As CModule
        Private m_CachedItems As CCollection(Of CacheItem)
        Private m_CacheSize As Integer
        Private m_AllItems As CCollection(Of T)
        Private m_isInit As Boolean
        Private m_Conn As CDBConnection = Nothing

        ''' <summary>
        ''' Oggetto utilizzabile per la sincronizzazione
        ''' </summary>
        ''' <remarks></remarks>
        Public ReadOnly cacheLock As New Object

        Protected Class CacheItem
            Public Item As T
            Public Live As Integer

            Public Sub New()
                DMDObject.IncreaseCounter(Me)
            End Sub

            Public Sub New(ByVal item As T)
                DMDObject.IncreaseCounter(Me)
                Me.Item = item
                Me.Live = 1
            End Sub

            Protected Overrides Sub Finalize()
                MyBase.Finalize()
                DMDObject.DecreaseCounter(Me)
            End Sub
        End Class


        ''' <summary>
        ''' Costruttore
        ''' </summary>
        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            Me.m_ModuleName = ""
            Me.m_CursorType = Nothing
            Me.m_Module = Nothing
            Me.m_CacheSize = 0
            Me.m_CachedItems = New CCollection(Of CacheItem)
            Me.m_AllItems = Nothing
            Me.m_isInit = False
        End Sub

        ''' <summary>
        ''' Costruttore
        ''' </summary>
        ''' <param name="moduleName"></param>
        Public Sub New(ByVal moduleName As String)
            Me.New()
            moduleName = Trim(moduleName)
            If (Left(moduleName, 3) <> "mod") Then Throw New ArgumentNullException("moduleName")
            If (moduleName = "") Then Throw New ArgumentNullException("moduleName")
            Me.m_ModuleName = moduleName
        End Sub

        ''' <summary>
        ''' Costruttore
        ''' </summary>
        ''' <param name="moduleName"></param>
        ''' <param name="cursorType"></param>
        Public Sub New(ByVal moduleName As String, ByVal cursorType As System.Type)
            Me.New()
            moduleName = Trim(moduleName)
            If (Left(moduleName, 3) <> "mod") Then Throw New ArgumentNullException("moduleName")
            If (moduleName = "") Then Throw New ArgumentNullException("moduleName")
            If (cursorType Is Nothing) Then Throw New ArgumentNullException("cursorType")
            Me.m_ModuleName = moduleName
            Me.m_CursorType = cursorType
        End Sub

        ''' <summary>
        ''' Costruttore
        ''' </summary>
        ''' <param name="moduleName"></param>
        ''' <param name="cursorType"></param>
        ''' <param name="cacheSize"></param>
        Public Sub New(ByVal moduleName As String, ByVal cursorType As System.Type, ByVal cacheSize As Integer)
            Me.New()
            moduleName = Strings.Trim(moduleName)
            If (Left(moduleName, 3) <> "mod") Then Throw New ArgumentNullException("moduleName")
            If (moduleName = "") Then Throw New ArgumentNullException("moduleName")
            If (cursorType Is Nothing) Then Throw New ArgumentNullException("cursorType")
            Me.m_ModuleName = moduleName
            Me.m_CursorType = cursorType
            Me.m_Module = Nothing
            Me.m_CacheSize = cacheSize
        End Sub

        ''' <summary>
        ''' Restituisce un riferimento al modulo
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property [Module] As CModule
            Get
                SyncLock Me
                    If (Me.m_Module Is Nothing) Then Me.m_Module = Sistema.Modules.GetItemByName(Me.m_ModuleName)
                    If (Me.m_Module Is Nothing) Then Me.m_Module = Me.CreateModuleInfo()
                    Return Me.m_Module
                End SyncLock
            End Get
        End Property

        ''' <summary>
        ''' Inizializza le risorse necessarie al modulo
        ''' </summary>
        Public Overridable Sub Initialize()
            If (Me.m_isInit) Then Return
            Dim m As CModule = Me.Module
            Me.m_isInit = True
        End Sub

        ''' <summary>
        ''' Rilascia le risorse necessarie al modulo
        ''' </summary>
        Public Overridable Sub Terminate()
            Me.m_isInit = False
        End Sub

        ''' <summary>
        ''' Crea il modulo e salva le informazioni nel DB
        ''' </summary>
        ''' <returns></returns>
        Protected Overridable Function CreateModuleInfo() As CModule
            Dim ret As New CModule(Me.m_ModuleName)
            ret.Stato = ObjectStatus.OBJECT_VALID
            ret.Description = Me.m_ModuleName
            ret.DisplayName = Me.m_ModuleName
            ret.ClassHandler = "CBaseModuleHandler"
            ret.Save()

            ret.InitializeStandardActions()
            Return ret
        End Function

        ''' <summary>
        ''' Crea un'istanza del cursore
        ''' </summary>
        ''' <returns></returns>
        Public Overridable Function CreateCursor() As DBObjectCursorBase
            Return Types.CreateInstance(Me.m_CursorType)
        End Function

        ''' <summary>
        ''' Restituisce l'elemento in base all'ID
        ''' </summary>
        ''' <param name="id"></param>
        ''' <returns></returns>
        Public Overridable Function GetItemById(ByVal id As Integer) As T
            SyncLock Me.cacheLock
                If (id = 0) Then Return Nothing
                Dim ret As T = Me.GetCachedItemById(id)
                If (ret Is Nothing) Then
                    ret = Me.GetNonCachedItemById(id)
                    If (ret IsNot Nothing) Then Me.AddToCache(ret)
                End If
                Return ret
            End SyncLock
        End Function

        Protected Overridable Function GetCachedItemById(ByVal id As Integer) As T
            If (id = 0) Then Return Nothing
            For i As Integer = 0 To Me.m_CachedItems.Count - 1
                Dim o As CacheItem = Me.m_CachedItems(i)
                If (GetID(o.Item) = id) Then Return o.Item
            Next
            Return Nothing
        End Function


        Public Overridable Function GetConnection() As CDBConnection
            SyncLock Me
                If (Me.m_Conn Is Nothing) Then
                    Dim cursor As DBObjectCursorBase = Me.CreateCursor
                    If (cursor IsNot Nothing) Then
                        Me.m_Conn = cursor.GetConnection
                        cursor.Dispose()
                    End If
                End If
                Return Me.m_Conn
            End SyncLock
        End Function


        Protected Overridable Function GetNonCachedItemById(ByVal id As Integer) As T
#If UseGetNonCachedItemById Then
            Dim conn As CDBConnection = Me.GetConnection
            Return conn.GetItemById(Me.Module, id)
#Else
            Dim dbRis As System.Data.IDataReader = Nothing

            Try
                If (id = 0) Then Return Nothing

                Dim item As T = Sistema.Types.CreateInstance(GetType(T))

                Dim conn As CDBConnection = DBUtils.GetConnection(item)

                Dim dbSQL As New System.Text.StringBuilder
                dbSQL.Append("SELECT * FROM [")
                dbSQL.Append(DBUtils.GetTableName(item))
                dbSQL.Append("] WHERE [ID]=")
                dbSQL.Append(DBUtils.DBNumber(id))

                dbRis = conn.ExecuteReader(dbSQL.ToString)

                If (dbRis.Read) Then
                    conn.Load(item, dbRis)
                Else
                    If (TypeOf (item) Is IDisposable) Then DirectCast(item, IDisposable).Dispose()
                    item = Nothing
                End If

                Return item
            Catch ex As Exception
                Sistema.Events.NotifyUnhandledException(ex)
                Throw
            Finally
                If (dbRis IsNot Nothing) Then dbRis.Dispose() : dbRis = Nothing
            End Try
#End If
        End Function

        Protected Overridable Sub AddToCache(ByVal item As T)
            If (Me.m_CacheSize = 0) Then Exit Sub
            Me.m_CachedItems.Add(New CacheItem(item))
            If (Me.m_CacheSize > 0) Then
                While (Me.m_CachedItems.Count > Me.m_CacheSize)
                    Me.m_CachedItems.RemoveAt(0)
                End While
            End If
            Me.m_AllItems = Nothing
        End Sub

        Protected Overridable Sub AddToCache1(ByVal item As T)
            If (Me.m_CacheSize = 0) Then Exit Sub
            Me.m_CachedItems.Add(New CacheItem(item))
            If (Me.m_CacheSize > 0) Then
                While (Me.m_CachedItems.Count > Me.m_CacheSize)
                    Me.m_CachedItems.RemoveAt(0)
                End While
            End If
        End Sub

        ''' <summary>
        ''' Aggiorna la cache sulla base dello stato dell'oggetto
        ''' </summary>
        ''' <param name="item"></param>
        ''' <remarks></remarks>
        Public Overridable Sub UpdateCached(ByVal item As T)
            SyncLock Me.cacheLock
                If (Me.m_CacheSize = 0) Then Exit Sub

                For i As Integer = 0 To Me.m_CachedItems.Count - 1
                    Dim o As CacheItem = Me.m_CachedItems(i)
                    If (GetID(o.Item) = GetID(item)) Then
                        Dim oldItem As T = o.Item
                        oldItem.InitializeFrom(item)
                        item = oldItem
                        Me.m_CachedItems.RemoveAt(i)
                        Exit For
                    End If
                Next


                If (TypeOf (item) Is IDBObject) Then
                    If (DirectCast(item, IDBObject).Stato = ObjectStatus.OBJECT_VALID) Then
                        Me.m_CachedItems.Add(New CacheItem(item))
                    End If
                ElseIf (GetID(item) <> 0) Then
                    Me.m_CachedItems.Add(New CacheItem(item))
                End If

                If (Me.m_CacheSize > 0) Then
                    While (Me.m_CachedItems.Count > Me.m_CacheSize)
                        Me.m_CachedItems.RemoveAt(0)
                    End While
                End If

                'If (Me.m_AllItems IsNot Nothing) Then
                '    For i As Integer = 0 To Me.m_AllItems.Count - 1
                '        If (GetID(Me.m_AllItems(i)) = GetID(item)) Then
                '            Dim oldItem As T = Me.m_AllItems(i)
                '            oldItem.InitializeFrom(item)
                '            Me.m_CachedItems.RemoveAt(i)
                '            Exit For
                '        End If
                '    Next
                'End If
                Me.m_AllItems = Nothing
            End SyncLock
        End Sub

        Public Overridable Sub RemoveCached(ByVal item As T)
            SyncLock Me.cacheLock
                If (Me.m_CacheSize = 0) Then Exit Sub
                For i As Integer = 0 To Me.m_CachedItems.Count - 1
                    Dim o As CacheItem = Me.m_CachedItems(i)
                    If (GetID(o.Item) = GetID(item)) Then
                        Me.m_CachedItems.RemoveAt(i)
                        Exit For
                    End If
                Next
                Me.m_AllItems = Nothing
            End SyncLock
        End Sub


        Public Overridable Function LoadAll() As CCollection(Of T)
            SyncLock Me.cacheLock
                If (Me.m_AllItems IsNot Nothing) Then Return Me.m_AllItems
                If (Me.m_CacheSize <> 0) Then
                    Dim cursor As DBObjectCursorBase = Nothing
#If Not DEBUG Then
                    try
#End If
                    cursor = Me.CreateCursor
                    cursor.PageSize = 10000
                    If (TypeOf (cursor) Is DBObjectCursor) Then
                        DirectCast(cursor, DBObjectCursor).Stato.Value = ObjectStatus.OBJECT_VALID
                    End If
                    cursor.IgnoreRights = True
                    While Not cursor.EOF AndAlso (Me.m_CacheSize < 0 OrElse Me.m_CachedItems.Count < Me.m_CacheSize)
                        Dim item1 As T = Me.GetCachedItemById(GetID(cursor.Item))
                        If (item1 Is Nothing) Then Me.AddToCache1(cursor.Item)
                        'Me.UpdateCached(cursor.Item)
                        cursor.MoveNext()
                    End While

#If Not DEBUG Then
                    Catch ex As Exception
                    throw
                    Finally
#End If
                    If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing
#If Not DEBUG Then
                    End Try
#End If
                End If

                Me.m_AllItems = New CCollection(Of T)
                For i As Integer = 0 To Me.m_CachedItems.Count - 1
                    Me.m_AllItems.Add(Me.m_CachedItems(i).Item)
                Next

                If (GetType(IComparable).IsAssignableFrom(GetType(T))) Then Me.m_AllItems.Sort()
                Return Me.m_AllItems
            End SyncLock
        End Function



        Protected ReadOnly Property CachedItems As CCollection(Of CacheItem)
            Get
                Return Me.m_CachedItems
            End Get
        End Property

        Protected Friend Sub doItemCreated(ByVal e As ItemEventArgs)
            RaiseEvent ItemCreated(Me, e)
        End Sub

        Protected Friend Sub doItemDeleted(ByVal e As ItemEventArgs)
            RaiseEvent ItemDeleted(Me, e)
        End Sub

        Protected Friend Sub doItemModified(ByVal e As ItemEventArgs)
            RaiseEvent ItemModified(Me, e)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class

End Class